### DISCUSSION TITLE
breaking the rules of hooks?

### DISCUSSION THREAD

**Discussion by maninkari** [statelyai/xstate#3108] [Q&A] [answered]
Hi, I am trying to write a widget (react component library) that connects to wallets and displays nfts. I am using x-state and I am wondering if I am getting some errors because of that when using different react versions.  

| widget   | app          | result  |
| -------- |----------| -------|
| react 17 | react 17 | all good |
| react 16 | react 17 | some linting issues |
| react 16 | react 16 | invalid hook call |
| react 17 | react 16 | Can't resolve '@xstate/react'|

Regarding row 3, could this be because of using the `send` hook within an event? something like:
``` 
wallet.on("connect", () => {
   if (wallet.connected) 
       send({ type: "CONNECT_WALLET" })
})
```
It works for version 17 though.

And on row 4 it can't resolve @xstate/react for some reason.
Does the xstate work with react 16.8 onwards? 

Thank you for any feedback and for writing and maintaining this xstate!

**Comment by davidkpiano** [comment]
Can you make a small reproducible example in CodeSandbox? You can use [one of the templates](https://github.com/statelyai/xstate#templates).

**Reply by maninkari** [reply]
"@darkblock.io/darkblock-widgets": "^0.0.9" -> in React v17.02
"@darkblock.io/darkblock-widgets": "^0.0.10" -> in React v16.14

App in react 17 here:
https://codesandbox.io/s/tender-haze-bpoqwv-bpoqwv?file=/src/Main/index.js

changing the versions leads to the different errors.

here's a fork with App in react 16:
https://codesandbox.io/s/tender-haze-bpoqwv-forked-wvzgkd

Haven't managed to reproduced the `Can't resolve '@xstate/react'` error with sandbox.
Thank you!

**Comment by Andarist** [comment, marked as answer]
As to the first codesandbox - you are importing there a component that is using XState machine. We can inspect its source code and check out that they are creating a machine in a React component through a factory and pass it to `useMachine`:
https://unpkg.com/browse/@darkblock.io/darkblock-widgets@0.0.10/dist/SolanaDarkblockWidget.js#L46

This machine should be memoized because we don't accept new machines at the moment - we can't "swap" them (like, we can't dispose the previous one and initialize the new one). This limitation might be lifted in the future - but right now it is what it is.

As to the second codesandbox - we can check out that `@solana/*` packages depend on React 17:
https://unpkg.com/browse/@solana/wallet-adapter-react@0.15.3/package.json#L33

And since you are using React 16 there - this won't work correctly. Usually React should be declared as a peer dep and this could potentially allow them to support `16.8.x || 17.x`.

Both issues are not quite problems with XState and should be addressed by the owners of those packages.

**Reply by maninkari** [reply]
Thank you so much for taking a look into this. I wrote the darkblock-widgets so I will memoize the stateMachine. 
Regarding the second codesandbox, one of our partners is using the wallet-adapter (maybe an older version written in react 16) with 16.13 and is requesting the darkblock-widget to be react 16.13 compatible. Glad to hear that it's not an xstate error and that @solana requiring react 17 is the main problem here.

**Reply by maninkari** [reply]
@Andarist @davidkpiano any suggestion on how to memoize the machine? It doesn't like this:

```
const Widget = ({
  tokenId = "",
  sessionToken = false,
  contractAddress = "",
  platform = "",
  wallets = null,
}) => {
  // have also tried this: const machine = useMemo(widgetMachine(tokenId, contractAddress, platform)); 
  const machine = useMemo(widgetMachine(tokenId, contractAddress, platform), [tokenId, contractAddress, platform]);
  const [state, send] = useMachine(machine);
```

Throws `Uncaught TypeError: nextCreate is not a function` and doesn't render anything

without the useMemo everything works but I sometimes get the following message when refreshing though: `Machine given to useMachine has changed between renders. This is not supported and might lead to unexpected results.
Please make sure that you pass the same Machine as argument each time.`

Is it a bad idea to use a state machine this way? the states are going to be the same for every platform and token (loading, connecting, decrypting, etc) but I have a state that needs to call some APIs and therefore I need to pass the parameters to the machine:

```
loading: {
        invoke: {
          src: () =>
            Promise.all([
              getOwner(contractAddress, tokenId, platform),
              getCreator(contractAddress, tokenId, platform),
              getData(tokenId, platform),
            ]),
          onDone: [
            {
```

Thank you!

**Reply by Andarist** [reply]
You are using the `useMemo` incorrectly. You need to return a value from a function, like this:
```js
const machine = useMemo(() => widgetMachine(tokenId, contractAddress, platform), []); 
```

However, in this situation it might even be better to do this since we accept machine factories too:
```js
const [state, send] = useMachine(() => widgetMachine(tokenId, contractAddress, platform));
```

**Reply by maninkari** [reply]
working like a charm, thank you!
